2020/day3.clj

```diff
@@ -7,27 +7,43 @@
        (string/split-lines)))
 
 (defn next-location
-  [loc [right down]]
-  [(+ (first loc) right)
-   (+ (second loc) down)])
+  [[row col] [right down]]
+  (let [next-row (+ row right)
+        next-col (+ col down)]
+    [next-row next-col]))
+
+(defn next-location-step
+  [right down n]
+  (next-location [0 0] [(* right n) (* down n)]))
 
 (defn locations-s
-  [height [right down]]
-  (take-while #(< (second %) height)
-              (map #(next-location [0 0] [(* right %) (* down %)])
-                   (range))))
+  [slope]
+  (let [right (first slope)
+        down  (second slope)]
+    (map (partial next-location-step right down)
+         (range))))
+
+(defn with-height?
+  [[_ col] height]
+  (< col height))
+
+(defn char-at-location
+  [grid [row col]]
+  (let [x (grid col)
+        y (mod row (count x))]
+    (nth x y)))
 
 (defn tree?
   [ch]
   (= ch \#))
 
 (defn count-trees-on-slope
-  [grid [right down]]
-  (let [height          (count grid)
-        valid-locations (locations-s height [right down])]
-    (->> valid-locations
-         (map #(get-in grid [(second %) (mod (first %)
-                                             (count (first grid)))]))
+  [grid slope]
+  (let [height    (count grid)
+        locations (locations-s slope)]
+    (->> locations
+         (take-while (fn [location] (with-height? location height)))
+         (map (fn [loc] (char-at-location grid loc)))
```

Comment:
I suggest pulling everything in the thread `->>` up to the `let` bindings so it can be named.

The only remaining core issue is that this code is filtering down the locations _before_ they are combined with the grid (`char-at-loc`). They need to be combined first, THEN you can take only the valid values

Reply:
Pulling the threading up to the bindings block does _not_ mean copy/pasting the full thread expression as a value in the let block. It means breaking down the threads and giving elements names. i.e. 
```
(-> a
   b
   c)
```
would instead be
```
(let [b-result (b a)
     c-result (c b)
     ...])
```
-------
With regard to the comment about incorrectly filtering location _before_ they are combined with the grid, the last change still has the same problem... nothing was changed.

I'll be more explicit : _the result of `map char-at-loc` should be an infinite sequence._